src/ruleset.hpp

```diff
@@ -123,6 +123,36 @@ struct ruleset {
         return root_addresses;
     }
 
+    [[nodiscard]] const std::vector<const char *> &get_available_action_types()
+    {
+        if (available_action_types.empty()) {
+            for (const auto &rule : rules) {
+                for (const auto &action : rule->get_actions()) {
+                    auto it = actions->find(action);
+                    if (it == actions->end()) {
+                        continue;
+                    }
+                    available_action_types_set.emplace(it->second.type_str);
+                }
+            }
+
+            for (const auto &[name, filter] : rule_filters) {
+                auto it = actions->find(filter->get_action());
+                if (it == actions->end()) {
+                    continue;
+                }
+                available_action_types_set.emplace(it->second.type_str);
+            }
+
+            // We generate everything at the end so in the future we can change
+            // the std::unordered_set for a flat_set.
```

Comment:
I don't think you need it at all, as a class member at least. You're only using it to reclaim back the memory, but afaics it's a self-inflicted problem (you don't need to copy `type_str`)